`spacetime init` rewrite

drogus opened the pull request:
This is a draft of the new functionality for `spacetime init`. In order to run it with built-in templates you have to set the path to the config file:

```
export SPACETIMEDB_CLI_TEMPLATES_FILE=crates/cli/.init-templates.json
```

In the future it will fetch the list from GH.

A few notes:

* the previous functionality of `spacetime init` does not work at the moment
* the code needs a bit more cleanup and tests before merging
* there is a bit of a mix in how we generate empty server and client projects. For Rust we use the existing way of generating. For TypeScript we clone an empty project from the repo. I wanted to play with both ways of doing things, and I'm still not sure which is better. Generation in Rust means that the generated code will match the CLI version and not necessarily whatever is in Git. On the other hand, for the builtin templates we will be fetching the newest version from GH, which I guess might also not what we want, ie. we probably want only stable templates. More discussion is needed here
* we use `spacetimedb` directory for the server files
* I don't particularly like the inability to disable interactive mode easily. We discussed disabling it by default if all of the required  arguments are passed, but I don't think it's feature proof. For example, if someone relies on a non-interactive mode, and we add a new required argument, instead of printing a message `missing --foo`, we will automatically launch interactive mode, which is harder to debug. That's why I think I'd prefer to implement `--non-interactive` argument
* it's kind of hard to keep the legacy behaviour. If you don't pass any arguments, we go into interactive mode. In the legacy version, we would print required arguments. If someone passes `--lang` or `--project-path` explicitly, I guess we could run the legacy workflow, but not sure if it's worth it, as the command was marked as unstable anyway
* the project path defaults to the project name, but I think we should probably replace change whitespaces to dashes, or at least ask for the project path with the project name being the default (or both)

drogus committed:
implement new spacetime init command

drogus committed:
Fix lints

drogus committed:
Format typescript code

drogus committed:
Update eslint config

drogus committed:
Small chnages in init args

drogus committed:
Fix smoke tests

drogus committed:
Fix cli refernece docs

drogus committed:
Bump versions to 1.6.*

drogus committed:
Merge remote-tracking branch 'origin/master' into drogus/spacetime-init

drogus committed:
Fix cli refernece docs

drogus committed:
Fix versions update

drogus committed:
Merge remote-tracking branch 'origin/master' into drogus/spacetime-init

drogus committed:
Move init/template to init, remove old functionality

drogus committed:
Add typescript chat templates to the list

drogus committed:
Use empty main.ts for epty typescript example

drogus committed:
lang choice and template are mutually exclusive

drogus committed:
Treat server and client lang as Option<>

drogus committed:
Add quickstart chat ts symlinks

drogus committed:
Update cli reference

drogus committed:
Change _BRANCH to _REFERENCE

drogus committed:
Fix error handling for unsupported langs

drogus committed:
Fix cloning a reference

drogus committed:
Merge remote-tracking branch 'origin/master' into drogus/spacetime-init

drogus committed:
Remove unneeded debug println

drogus commented:
This should be now read to review. The way it should work is to go into interactive mode right away, and ask for options, for example:

<img width="2540" height="650" alt="image" src="https://github.com/user-attachments/assets/c30bfbee-30d8-4fc7-946f-ae87845bf482" />

If a user passes all the required options or `--non-interactive`, the command will not go into interactive mode.

drogus committed:
Merge interactive and non-interactive functions

drogus committed:
Install dependencies for a typescript server

drogus committed:
Include built-in templates in the binary

cloutiertyler commented:
The following tests gave me an error:

<img width="2074" height="738" alt="image" src="https://github.com/user-attachments/assets/77ec744c-2c1a-4884-a737-412491480199" />
<img width="2090" height="578" alt="image" src="https://github.com/user-attachments/assets/f0de2288-21e3-4268-8156-79e624ef66c5" />

drogus committed:
Panic if a template has 0 files

drogus committed:
Bring back the 'none' option on init's clients list

drogus committed:
refactor cli's build.rs and add more comments

drogus committed:
Remove client-lang option from init

It was added first when the requirements were different. At the moment
we only allow to chose server-lang if template is not provided

drogus committed:
Add missing templates list

drogus committed:
remove unused functions

drogus committed:
Fix cli reference

drogus committed:
Fix lints

drogus committed:
update cargo.lock

drogus committed:
Fix init in smoketests

drogus committed:
Fix reference so spacetime init

drogus committed:
Update the name of both the server and the client for known languages

drogus committed:
Fix checking for login in non-interactive mode

drogus committed:
Try updatring cargo.lock manually

drogus committed:
Fix smoketests

drogus committed:
Change bitflags version

drogus committed:
Try to fix smoke tests

cloutiertyler committed:
Small fixes and changes

drogus committed:
Don't use symlinks for templates and copy files into the crate when building

drogus committed:
Merge remote-tracking branch 'origin/drogus/spacetime-init' into drogus/spacetime-init

drogus committed:
add basic-c-sharp client

drogus committed:
Exclude spacetimedb directory from c# clients

cloutiertyler committed:
Small fix for windows

cloutiertyler committed:
Fixes a bug with project path on typescript build

cloutiertyler committed:
Fixed several bugs on windows

cloutiertyler committed:
Merge branch 'master' into drogus/spacetime-init

Signed-off-by: Tyler Cloutier <cloutiertyler@users.noreply.github.com>

drogus committed:
Change bitflags version to 2.10.0

cloutiertyler commented on docs/docs/modules/c-sharp/index.md:
````diff
@@ -66,7 +66,7 @@ Tables and reducers in C# modules can use any type annotated with [`[SpacetimeDB
 
 ```bash
-spacetime init --lang csharp my-project-directory
+spacetime init --non-interactive --name my-spacetimedb-project --server-lang csharp my-project-directory
````
Several suggestions:

1. `--name my-spacetimedb-project my-spacetimedb-directory` is a little redundant, let's make it so that the parameter is called `--project-path` instead and that parameter tells you where you're going put the project. Let's make the positional argument the project name (more in line with what `spacetime publish` does). Let's make it so that if you do not specify a project path, it creates a directory in the current directory with the same name as the project name.
2. Let's leave off `--non-interactive` in the docs, so it might actually go into interactive mode for some stuff.
3. Let's rename `--server-lang` to `--lang` (I'm less sure about this change. I would be okay to keep it `--server-lang` if we made the other changes).

So this line would become:

```bash
spacetime init --lang csharp my-spacetimedb-project
```

jdetter commented on crates/cli/src/subcommands/init.rs:
```diff
@@ -1,28 +1,979 @@
+            template_type: TemplateType::GitHub,
+            server_lang: None,
+            client_lang: None,
+            github_repo: Some(template_str.to_string()),
```
Should we check to see if this looks like a URL here? Like if someone just mistypes a template name we will end up here too no?

jdetter reviewed (commented):
Some UX questions/nits about github templates, I'm still reviewing

jdetter commented on crates/cli/src/subcommands/init.rs:
```diff
@@ -1,28 +1,979 @@
+    }
+}
+
+fn clone_github_template(repo_input: &str, target: &Path) -> anyhow::Result<()> {
```
I think having them part of SpacetimeDB is a good idea. So then you should be able to do something like this then right?

```
spacetime init --template https://github.com/clockworklabs/SpacetimeDB/templates/Blackholio/Unity
```

jdetter commented on crates/cli/src/subcommands/init.rs:
```diff
@@ -1,28 +1,979 @@
+            println!("  {} - {}", template.id, template.description);
+        }
+        println!();
+        loop {
```
This loop isn't used

jdetter reviewed (commented):
More review

jdetter commented on crates/cli/src/subcommands/init.rs:
```diff
@@ -1,28 +1,979 @@
+            1 => Some("pnpm"),
+            2 => Some("yarn"),
+            3 => Some("bun"),
+            _ => None,
```
Let's remove the "other" option then?

drogus replied:
If we remove the "other" option you would be forced to choose one of the package managers, or am I missing something?

drogus committed:
Rename --server-lang back to --lang

drogus committed:
Copy template files in build.rs only if the file changed

drogus committed:
Update edition and packages referencing workspace in generated rust code

drogus committed:
Make init's project name a positional arguments and project path a named argument

drogus committed:
Merge branch 'drogus/spacetime-init' of github.com:clockworklabs/SpacetimeDB into drogus/spacetime-init

cloutiertyler committed:
Fixed small issues and warnings

cloutiertyler committed:
Merged master

cloutiertyler committed:
Updated all the template projects so that that use environment variables with fallbacks for connecting to SpacetimeDB

cloutiertyler committed:
Fixed pm command on windows

cloutiertyler committed:
pnpm format

cloutiertyler commented on crates/cli/src/subcommands/init.rs:
```diff
@@ -1,28 +1,979 @@
+        format!("https://github.com/{}", repo_input)
+    } else {
+        anyhow::bail!("Invalid repository format. Use 'owner/repo' or full URL");
+    };
```
I have made this check more robust:

```rs
    let is_git_url = |s: &str| {
        s.starts_with("git@") || s.starts_with("ssh://") || s.starts_with("http://") || s.starts_with("https://")
    };

    let repo_url = if is_git_url(repo_input) {
        repo_input.to_string()
    } else if repo_input.contains('/') {
        format!("https://github.com/{}", repo_input)
    } else {
        anyhow::bail!("Invalid repository format. Use 'owner/repo' or full git clone URL");
    };
```

cloutiertyler committed:
Addresses John's comments on the PR

drogus committed:
Move parsing TS SDK version to build.rs to not include_str a file outside of the crate

jdetter commented on docs/docs/03-Unity Tutorial/03-part-2.md:
````diff
@@ -23,7 +23,7 @@ In your `blackholio` directory, run the following command to initialize the Spac
 
 ```bash
-spacetime init --lang=rust server-rust
+spacetime init --lang rust server-rust
````
In this case I'm being prompted for a project path, I think most users would be smart enough to just press enter to do the default thing but maybe it would be good to just include the `--project-path` option here so that users don't get prompted?

jdetter reviewed (commented):
Some docs feedback for now

cloutiertyler committed:
Hopefully updated C# tests and now updating version for the C# built-in templates

cloutiertyler committed:
Fixed compile error

cloutiertyler committed:
Added --server-only flag

jdetter committed:
Initial pass on updating the docs for the new `spacetime init`
functionality

jdetter committed:
Merge branch 'drogus/spacetime-init' of ssh://github.com/clockworklabs/SpacetimeDB into drogus/spacetime-init

jdetter committed:
Unity docs clarification

jdetter committed:
Changes required for unreal engine - still needs testing

cloutiertyler committed:
Update quickstarts for new spacetime init command

jdetter committed:
Finished testing the Unity tutorial + fixed the `--server-only` bug

jdetter committed:
Both tutorials are working

cloutiertyler committed:
npm install on both server and client if client is also typescript

cloutiertyler committed:
cargo fmt

jdetter committed:
Testing change

jdetter committed:
Small update

jdetter committed:
Small fix

jdetter committed:
Add exclude for spacetimedb by default

jdetter committed:
Merge branch 'master' into drogus/spacetime-init

Signed-off-by: John Detter <4099508+jdetter@users.noreply.github.com>

jdetter committed:
Fixes for the quickstarts so far

jdetter committed:
Merge branch 'drogus/spacetime-init' of ssh://github.com/clockworklabs/SpacetimeDB into drogus/spacetime-init

cloutiertyler committed:
Fixed small .env bug in init

cloutiertyler committed:
cargo fmt

jdetter committed:
Fix docs

jdetter committed:
Merge branch 'drogus/spacetime-init' of ssh://github.com/clockworklabs/SpacetimeDB into drogus/spacetime-init

jdetter reviewed (approved):
I've tested 
- [x] Fixed the --server-only bug which prevented generating the module code
- [x] Finished testing the Unity tutorial for both rust and C# modules up to the spacetime generate ... part
- [x] Test Unreal tutorial with both C# and rust up to the spacetime generate ... part
- [x] Fix the issue with Cargo.toml in the rust quickstart
- [x] Test rust module quickstart
- [x] Test C# module quickstart
- [ ] Test typescript module quickstart

There are a few code things that I would probably change but I think we can do those as follow-up PRs

cloutiertyler reviewed (approved):
Okay, John and I have thoroughly tested and fixed small bugs. There are two remaining items that are pretty gnarly in this version:

1. We scoop up the slop from all over the repo, and we realllllly should change that to be a straightforward copy of future `/templates` directory
2. We need a better gameplan for ensuring that the templates locally depend on SpacetimeDB if you have the repo checked out, but use a package manager/versioning scheme if you copy them as a template. It's pretty slapdash right now, but it kinda works. We should make it better (standardized).

I'm okay to do these things later.
